classes/client/do_client.php

```diff
@@ -0,0 +1,128 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * DO client.
+ *
+ * @package   tool_objectfs
+ * @author    Brian Yanosik <kisonay@gmail.com>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_objectfs\client;
+
+defined('MOODLE_INTERNAL') || die();
+
+$autoloader = $CFG->dirroot . '/local/aws/sdk/aws-autoloader.php';
+
+if (!file_exists($autoloader)) {
+
+    // Stub class with bare implementation for when the SDK prerequisite does not exist.
+    class do_client {
+        public function get_availability() {
+            return false;
+        }
+
+        public function register_stream_wrapper() {
+            return false;
+        }
+    }
+
+    return;
+}
+
+require_once($autoloader);
+
+use Aws\S3\S3Client;
+
+class do_client extends s3_client {
+
+    public function __construct($config) {
+        $this->bucket = $config->do_space;
+        $this->set_client($config);
+    }
+
+    public function set_client($config) {
+        $this->client = S3Client::factory(array(
+            'credentials' => array('key' => $config->do_key, 'secret' => $config->do_secret),
+            'region' => $config->do_region,
+            'endpoint' => 'https://' . $config->do_region . '.digitaloceanspaces.com',
```

Comment:
Do you think "'.digitaloceanspaces.com" could change in the future?

Reply:
Anything is possible but there would have to be an announcement/notice on their part as it would be an API change.

Reply:
So  you recon it's an overkill to make it configurable?

Reply:
Anything is possible but they would have to make an announcement/provide notice on their part as it would be an API change.

Reply:
I feel like it would be but at the same time, probably wouldn't hurt for it to be configurable. where would you want that defined? lib.php?

Reply:
You could probably replace do_region with just endpoint and then ask the user to submit the whole endpoint including "https://"

Reply:
If you feel like you want to add this, then the best place should be a config form.

Reply:
I'll go ahead and make this change.

Reply:
> You could probably replace do_region with just endpoint and then ask the user to submit the whole endpoint including "https://"

Not sure if I like this. I'd prefer to limit a user to existing regions as it's done now.

Reply:
less room for user configuration error by limiting it to a dropdown like it is now but will require maintenance if DigitalOcean decides to add regions or change things on their end.

Reply:
yeah... Anyway up to you :+1:

Reply:
I like it how it is (I had originally contemplated making it more open-ended when I was making it S3 compatible generic) but once I made it specific to DigitalOcean I decided to make it more specific.

Reply:
I don't think it's a big deal either ways, so feel free to make a decision yourself. :)

Reply:
Let's keep it as is. As I will be using this regularly, ill be keeping on top of any region changes etc. If you require any other changes, let me know.